xarray/backends/rasterio_.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+
+try:
+    import rasterio
+except ImportError:
+    rasterio = False
+
+from .. import Variable, DataArray
+from ..core.utils import FrozenOrderedDict, Frozen, NDArrayMixin
+from ..core import indexing
+from ..core.pycompat import OrderedDict, suppress
+
+from .common import AbstractDataStore
+
+__rio_varname__ = 'raster'
+
+
+class RasterioArrayWrapper(NDArrayMixin):
+    def __init__(self, ds):
+        self._ds = ds
+        self.array = ds.read()
+
+    @property
+    def dtype(self):
+        return np.dtype(self._ds.dtypes[0])
+
+    def __getitem__(self, key):
+        if key == () and self.ndim == 0:
+            return self.array.get_value()
+        return self.array[key]
```

Comment:
See also: https://gist.github.com/lpinner/bd57b54a5c6903e4a6a2 (can't reuse this directly, though, because it doesn't have a license).

Anyways, I would definitely see if the rasterio folks are up for implementing a `__getitem__` method. There's no reason why this is xarray specific, to be honest.